velox/functions/prestosql/DateTimeFunctions.h

```diff
@@ -1392,4 +1397,30 @@ struct TimeZoneMinuteFunction : public TimestampWithTimezoneSupport<T> {
   }
 };
 
+template <typename T>
+struct ToISO8601FromTimestampWithTimezoneFunction
+    : public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<TimestampWithTimezone>& timestampWithTimezone) {
+    auto timestamp = unpackTimestampUtc(timestampWithTimezone);
+    auto timezoneID = unpackZoneKeyId(timestampWithTimezone);
+    auto timezoneName = util::getTimeZoneName(timezoneID);
+
+    timestamp.toTimezone(timezoneID);
+    auto tsStr = timestamp.toString();
+
+    if (isalpha(timezoneName[0])) {
+      auto starting_timepoint = date::make_zoned(
+          timezoneName,
+          date::sys_seconds{(std::chrono::seconds)(timestamp.getSeconds())});
+      timezoneName = format("%Ez", starting_timepoint);
+    }
+
+    result = tsStr.substr(0, 19) + ".000" + timezoneName;
```

Comment:
The timestamp with timezone stores the milliseconds. So you have to account for the subsecond precision and cannot ignore it here by hard coding it.
The timestamp extracted here may have the subscond precision set and it needs to be reflected when the formatting occurs.

Reply:
Thanks for catching this - I forgot to update the rounding/millisecond formatting logic here after I did so on the Timestamp PR. Fixed and added tests - though rounding up is not going to work yet due to the absence of the changes I made to `Timestamp.cpp` in that PR. To avoid conflicts, I think I'll just leave the implementation there and let it land first.

Reply:
Fixed and added tests with millisecond values.